README
ocaml-fdkaac

This package contains an OCaml interface for 
the fdk=aac library.

Please read the COPYING file before using this software.

Prerequisites:
==============

- ocaml >= 4.00.1 (haven't tried earlier versions)

- fdk-aac >= 0.1.1

- findlib >= 0.8.1 (haven't tried earlier versions)

Compilation:
============

	$ make all

This should build both the native and the byte-code version of the
extension library.

Installation:
=============

	$ make install

This should install the library file (using ocamlfind) in the
appropriate place.

Author:
=======

This author of this software may be contacted by electronic mail
at the following address: savonet-users@lists.sourceforge.net.


CHANGES
0.3.3 (04-01-2022)
=====
* Bring back dynamic loader (#5)

0.3.2 (27-03-2020)
=====
* Compatibility with version 2.0.0.
* Switch to dune for building.

0.3.1 (22-09-2019)
=====
* Check for fdk-aac version <= 0.1.6

0.3.0 (02-08-2019)
=====
* Fixed type for `Bandwidth param.

0.2.1 (03-08-2015)
=====
* Added variable bitrate mode.

0.2.0 (18-05-2013)
=====
* Added dynamic loader

0.1.0 (08-05-2013)
=====
* Initial release


src/fdkaac_stubs.c
/*
 * OCaml bindings for libfdk-aac
 *
 * Copyright 2013 Savonet team
 *
 * This file is part of ocaml-fdk-aac.
 *
 * ocaml-fdk-aac is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * ocaml-fdk-aac is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with ocaml-fdk-aac; if not, write to the Free Software
 * Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
 */

/* OCaml bindings for the libfdk-aac library. */

#include <caml/alloc.h>
#include <caml/callback.h>
#include <caml/custom.h>
#include <caml/fail.h>
#include <caml/memory.h>
#include <caml/misc.h>
#include <caml/signals.h>

#include <fdk-aac/aacenc_lib.h>

#include <stdint.h>
#include <string.h>

/* Not all errors seem to be used
 * so we only check the one that
 * appear relevant.. */
static void check_for_err(int ret) {
  switch (ret) {
  case AACENC_OK:
    return;

  case AACENC_INVALID_HANDLE:
    caml_raise_constant(*caml_named_value("fdkaac_exn_invalid_handle"));
    break;

  case AACENC_MEMORY_ERROR:
    caml_raise_out_of_memory();
    break;

  case AACENC_UNSUPPORTED_PARAMETER:
    caml_raise_constant(*caml_named_value("fdkaac_exn_unsupported_parameter"));
    break;

  case AACENC_INVALID_CONFIG:
    caml_raise_constant(*caml_named_value("fdkaac_exn_invalid_config"));
    break;

  case AACENC_INIT_ERROR:
    caml_raise_with_arg(*caml_named_value("fdkaac_exn_error"), Val_int(0));
    break;

  case AACENC_INIT_AAC_ERROR:
    caml_raise_with_arg(*caml_named_value("fdkaac_exn_error"), Val_int(1));
    break;

  case AACENC_INIT_SBR_ERROR:
    caml_raise_with_arg(*caml_named_value("fdkaac_exn_error"), Val_int(2));
    break;

  case AACENC_INIT_TP_ERROR:
    caml_raise_with_arg(*caml_named_value("fdkaac_exn_error"), Val_int(3));
    break;

  case AACENC_INIT_META_ERROR:
    caml_raise_with_arg(*caml_named_value("fdkaac_exn_error"), Val_int(4));
    break;

  case AACENC_ENCODE_ERROR:
    caml_raise_with_arg(*caml_named_value("fdkaac_exn_error"), Val_int(5));
    break;

  case AACENC_ENCODE_EOF:
    caml_raise_constant(*caml_named_value("fdkaac_exn_encode_eof"));
    break;

  default:
    caml_raise_with_arg(*caml_named_value("fdkaac_exn_unknown_error"),
                        Val_int(ret));
    break;
  }
}

#define Encoder_val(v) (*((HANDLE_AACENCODER *)Data_custom_val(v)))

static void finalize_encoder(value e) {
  HANDLE_AACENCODER enc = Encoder_val(e);
  aacEncClose(&enc);
}

static struct custom_operations encoder_ops = {
    "ocaml_fdkaac_encoder",   finalize_encoder,
    custom_compare_default,   custom_hash_default,
    custom_serialize_default, custom_deserialize_default};

#include <stdio.h>

CAMLprim value ocaml_fdkaac_init_enc(value chans) {
  CAMLparam0();
  CAMLlocal1(ans);

  int channels = Int_val(chans);
  int mode;

  HANDLE_AACENCODER enc;

  int ret = aacEncOpen(&enc, 0x0, channels);
  if (ret != AACENC_OK) {
    free(enc);
    check_for_err(ret);
  }

  switch (channels) {
  case 1:
    mode = MODE_1;
    break;
  case 2:
    mode = MODE_2;
    break;
  case 3:
    mode = MODE_1_2;
    break;
  case 4:
    mode = MODE_1_2_1;
    break;
  case 5:
    mode = MODE_1_2_2;
    break;
  case 6:
    mode = MODE_1_2_2_1;
    break;
  default:
    free(enc);
    caml_raise_constant(*caml_named_value("fdkaac_exn_unsupported_parameter"));
  }

  ret = aacEncoder_SetParam(enc, AACENC_CHANNELMODE, mode);

  if (ret != AACENC_OK) {
    free(enc);
    check_for_err(ret);
  }

  ret = aacEncoder_SetParam(enc, AACENC_CHANNELORDER, 1);

  if (ret != AACENC_OK) {
    free(enc);
    check_for_err(ret);
  }

  ans = caml_alloc_custom(&encoder_ops, sizeof(HANDLE_AACENCODER), 1, 0);
  Encoder_val(ans) = enc;

  CAMLreturn(ans);
}

CAMLprim value ocaml_fdkaac_encode(value e, value buf, value ofs, value len) {
  CAMLparam2(e, buf);
  CAMLlocal1(ret);

  HANDLE_AACENCODER enc = Encoder_val(e);

  AACENC_BufDesc inBuf = {0};
  AACENC_BufDesc outBuf = {0};
  AACENC_InArgs inArgs = {0};
  AACENC_OutArgs outArgs = {0};

  // Input buffer
  int bufid = IN_AUDIO_DATA;
  int buflen = Int_val(len);
  int bufelsize = 2; // S16LE
  int offset = Int_val(ofs);
  void *bufs;

  inBuf.numBufs = 1;
  inBuf.bufferIdentifiers = &bufid;
  inBuf.bufSizes = &buflen;
  inBuf.bufElSizes = &bufelsize;

  bufs = malloc(buflen);

  if (bufs == NULL)
    caml_raise_out_of_memory();

  memcpy(bufs, String_val(buf) + offset, buflen);

  inBuf.bufs = &bufs;

  inArgs.numInSamples = buflen / bufelsize;

  // Output buffer
  uint8_t outbuf[20480];
  int outid = OUT_BITSTREAM_DATA;
  int outlen = sizeof(outbuf);
  int outelsize = 1;
  void *outptr = &outbuf;

  outBuf.numBufs = 1;
  outBuf.bufferIdentifiers = &outid;
  outBuf.bufSizes = &outlen;
  outBuf.bufElSizes = &outelsize;
  outBuf.bufs = &outptr;

  caml_enter_blocking_section();
  int err = aacEncEncode(enc, &inBuf, &outBuf, &inArgs, &outArgs);
  caml_leave_blocking_section();

  free(bufs);
  check_for_err(err);

  ret = caml_alloc_string(outArgs.numOutBytes);
  memcpy(Bp_val(ret), outbuf, outArgs.numOutBytes);

  CAMLreturn(ret);
}

CAMLprim value ocaml_fdkaac_flush(value e) {
  CAMLparam1(e);
  CAMLlocal1(ret);

  HANDLE_AACENCODER enc = Encoder_val(e);

  AACENC_BufDesc inBuf = {0};
  AACENC_BufDesc outBuf = {0};
  AACENC_InArgs inArgs = {0};
  AACENC_OutArgs outArgs = {0};

  inArgs.numInSamples = -1;

  // Output buffer
  uint8_t outbuf[20480];
  int outid = OUT_BITSTREAM_DATA;
  int outlen = sizeof(outbuf);
  int outelsize = 1;
  void *outptr = &outbuf;

  outBuf.bufferIdentifiers = &outid;
  outBuf.bufSizes = &outlen;
  outBuf.bufElSizes = &outelsize;
  outBuf.bufs = &outptr;

  caml_enter_blocking_section();
  int err = aacEncEncode(enc, &inBuf, &outBuf, &inArgs, &outArgs);
  caml_leave_blocking_section();

  check_for_err(err);

  ret = caml_alloc_string(outArgs.numOutBytes);
  memcpy(Bp_val(ret), outbuf, outArgs.numOutBytes);

  CAMLreturn(ret);
}

CAMLprim value ocaml_fdkaac_set_param(value e, value p, value v) {
  CAMLparam3(e, p, v);

  HANDLE_AACENCODER enc = Encoder_val(e);

  AACENC_PARAM param = Int_val(p);

  check_for_err(aacEncoder_SetParam(enc, param, Int_val(v)));

  CAMLreturn(Val_unit);
}

CAMLprim value ocaml_fdkaac_get_param(value e, value p) {
  CAMLparam2(e, p);
  int ans;

  HANDLE_AACENCODER enc = Encoder_val(e);

  AACENC_PARAM param = Int_val(p);

  ans = aacEncoder_GetParam(enc, param);

  CAMLreturn(Val_int(ans));
}

CAMLprim value ocaml_fdkaac_info(value e) {
  CAMLparam1(e);
  CAMLlocal1(ans);
  HANDLE_AACENCODER enc = Encoder_val(e);
  AACENC_InfoStruct info = {0};

  // Apparently we need to do this before being able to access info...
  check_for_err(aacEncEncode(enc, NULL, NULL, NULL, NULL));
  check_for_err(aacEncInfo(enc, &info));

  ans = caml_alloc_tuple(2);
  Store_field(ans, 0, Val_int(info.inputChannels));
  Store_field(ans, 1, Val_int(info.frameLength));
  CAMLreturn(ans);
}


examples/wav2aac.ml
(*
 * Copyright 2013 Savonet team
 *
 * This file is part of OCaml-Vorbis.
 *
 * OCaml-Vorbis is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * OCaml-Vorbis is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with OCaml-Vorbis; if not, write to the Free Software
 * Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
 *)

let src = ref ""
let dst = ref ""

let input_string chan len =
  let ans = Bytes.create len in
  (* TODO: check length *)
  ignore (input chan ans 0 len);
  Bytes.to_string ans

let input_int chan =
  let buf = input_string chan 4 in
  int_of_char buf.[0]
  + (int_of_char buf.[1] lsl 8)
  + (int_of_char buf.[2] lsl 16)
  + (int_of_char buf.[3] lsl 24)

let input_short chan =
  let buf = input_string chan 2 in
  int_of_char buf.[0] + (int_of_char buf.[1] lsl 8)

let vbr = ref 0
let bitrate = ref 64000

let aot_of_string = function
  | "MPEG4 AAC-LC" -> `Mpeg_4 `AAC_LC
  | "MPEG4 HE-AAC" -> `Mpeg_4 `HE_AAC
  | "MPEG4 HE-AAC v2" -> `Mpeg_4 `HE_AAC_v2
  | "MPEG4 AAC-LD" -> `Mpeg_4 `AAC_LD
  | "MPEG4 AAC-ELD" -> `Mpeg_4 `AAC_ELD
  | "MPEG2 AAC-LC" -> `Mpeg_2 `AAC_LC
  | "MPEG2 HE-AAC" -> `Mpeg_2 `HE_AAC
  | "MPEG2 HE-AAC v2" -> `Mpeg_2 `HE_AAC_v2
  | _ -> raise Fdkaac.Encoder.Unsupported_parameter

let string_of_aot = function
  | `Mpeg_4 `AAC_LC -> "MPEG4 AAC-LC"
  | `Mpeg_4 `HE_AAC -> "MPEG4 HE-AAC"
  | `Mpeg_4 `HE_AAC_v2 -> "MPEG4 HE-AAC v2"
  | `Mpeg_4 `AAC_LD -> "MPEG4 AAC-LD"
  | `Mpeg_4 `AAC_ELD -> "MPEG4 AAC-ELD"
  | `Mpeg_2 `AAC_LC -> "MPEG2 AAC-LC"
  | `Mpeg_2 `HE_AAC -> "MPEG2 HE-AAC"
  | `Mpeg_2 `HE_AAC_v2 -> "MPEG2 HE-AAC v2"

let aots =
  [
    "MPEG4 AAC-LC";
    "MPEG4 HE-AAC";
    "MPEG4 HE-AAC v2";
    "MPEG4 AAC-LD";
    "MPEG4 AAC-ELD";
    "MPEG2 AAC-LC";
    "MPEG2 HE-AAC";
    "MPEG42 HE-AAC v2";
  ]

let aot = ref (`Mpeg_4 `HE_AAC_v2)
let afterburner = ref true
let sbr = ref true
let usage = "usage: wav2aac [options] source destination"

let _ =
  let aots = List.map (Printf.sprintf "%S") aots in
  Arg.parse
    [
      ("--vbr", Arg.Int (fun i -> vbr := i), "Encode in variable bitrate.");
      ( "--bitrate",
        Arg.Int (fun b -> bitrate := b * 1000),
        "Bitrate, in kilobits per second, defaults to 64kbps" );
      ("--afterburner", Arg.Set afterburner, "Enable afterburner effect.");
      ( "--aot",
        Arg.String (fun x -> aot := aot_of_string x),
        Printf.sprintf "Audio object type. Possible values:\n    %s"
          (String.concat ",\n    " aots) );
      ("--no-sbr", Arg.Clear sbr, "Disable spectral band replication.");
    ]
    (let pnum = ref (-1) in
     fun s ->
       incr pnum;
       match !pnum with
         | 0 -> src := s
         | 1 -> dst := s
         | _ ->
             Printf.eprintf "Error: too many arguments\n";
             exit 1)
    usage;
  if !src = "" || !dst = "" then (
    Printf.printf "%s\n" usage;
    exit 1);
  let ic = open_in_bin !src in
  let oc = open_out_bin !dst in
  (* TODO: improve! *)
  if input_string ic 4 <> "RIFF" then invalid_arg "No RIFF tag";
  ignore (input_string ic 4);
  if input_string ic 4 <> "WAVE" then invalid_arg "No WAVE tag";
  if input_string ic 4 <> "fmt " then invalid_arg "No fmt tag";
  let _ = input_int ic in
  let _ = input_short ic in
  (* TODO: should be 1 *)
  let channels = input_short ic in
  let infreq = input_int ic in
  let _ = input_int ic in
  (* bytes / s *)
  let _ = input_short ic in
  (* block align *)
  let bits = input_short ic in
  let enc = Fdkaac.Encoder.create channels in
  Fdkaac.Encoder.set enc (`Aot !aot);
  if !vbr <> 0 then Fdkaac.Encoder.set enc (`Bitrate_mode (`Variable !vbr))
  else begin
    Fdkaac.Encoder.set enc (`Bitrate_mode `Constant);
    Fdkaac.Encoder.set enc (`Bitrate !bitrate)
  end;
  Fdkaac.Encoder.set enc (`Transmux `Adts);
  Fdkaac.Encoder.set enc (`Samplerate infreq);
  Fdkaac.Encoder.set enc (`Afterburner !afterburner);
  Fdkaac.Encoder.set enc (`Sbr_mode !sbr);
  let buflen = 1024 in
  let data = Bytes.create buflen in
  let start = Unix.time () in
  Printf.printf "Input detected: PCM WAVE %d channels, %d Hz, %d bits\n%!"
    channels infreq bits;
  let br_info =
    if !vbr <> 0 then Printf.sprintf "VBR %d" !vbr
    else Printf.sprintf "%d kbps" (!bitrate / 1000)
  in
  Printf.printf
    "Encoding to: %s, %d channels, %d Hz, %s, afterburner: %b, sbr: %b\n\
     Please wait...\n\
     %!"
    (string_of_aot !aot) channels infreq br_info !afterburner !sbr;
  (* skip headers *)
  let rec aux () =
    let tag = input_string ic 4 in
    match tag with
      | "LIST" ->
          let n = input_int ic in
          let _ = input_string ic n in
          aux ()
      | "data" -> ()
      | _ -> invalid_arg "No data tag"
  in
  aux ();
  begin
    try
      while true do
        let len = input ic data 0 buflen in
        let data = Bytes.to_string data in
        if len = 0 then raise End_of_file;
        let ret = Fdkaac.Encoder.encode enc data 0 len in
        output_string oc ret
      done
    with
      | End_of_file -> ()
      | Fdkaac.Encoder.Error _ as e ->
          failwith
            (match Fdkaac.Encoder.string_of_exception e with
              | Some s -> s
              | None -> "Unknown error.")
  end;
  let ret =
    try Fdkaac.Encoder.flush enc
    with Fdkaac.Encoder.Unsupported_parameter ->
      Printf.printf "Could not flush.\n";
      ""
  in
  output_string oc ret;
  close_in ic;
  close_out oc;
  Printf.printf "Finished in %.0f seconds.\n" (Unix.time () -. start);
  Gc.full_major ()


src/fdkaac_loader.ml
(*
 * Copyright 2013 Savonet team
 *
 * This file is part of ocaml-fdkaac.
 *
 * ocaml-fdkaac is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * ocaml-fdkaac is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with ocaml-fdkaac; if not, write to the Free Software
 * Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
 *)

(* $Id$ *)

(**
  * Bindings to the fdkaac AAC library.
  * Dynamic loading module.
  *
  * @author Romain Beauxis
  *)

open Fdkaac_dynlink

let () = handler.fdkaac_module <- Some (module Fdkaac : Fdkaac_t)


src/fdkaac.ml
(*
 * Copyright 2013 Savonet team
 *
 * This file is part of ocaml-fdkaac.
 *
 * ocaml-fdkaac is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * ocaml-fdkaac is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with ocaml-fdkaac; if not, write to the Free Software
 * Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
 *)

(** OCaml bindings for the libfdk-aac. *)

module Encoder = struct
  exception Invalid_handle
  exception Unsupported_parameter
  exception Invalid_config
  exception Invalid_data
  exception Error of int
  exception End_of_file
  exception Unknown of int

  let () =
    Callback.register_exception "fdkaac_exn_invalid_handle" Invalid_handle;
    Callback.register_exception "fdkaac_exn_unsupported_parameter"
      Unsupported_parameter;
    Callback.register_exception "fdkaac_exn_invalid_config" Invalid_config;
    Callback.register_exception "fdkaac_exn_error" (Error 0);
    Callback.register_exception "fdkaac_exn_encode_eof" End_of_file;
    Callback.register_exception "aacplus_exn_unknown_error" (Unknown 0)

  let string_of_exception = function
    | Invalid_handle -> Some "Handle passed to function call was invalid"
    | Unsupported_parameter -> Some "Parameter not available"
    | Invalid_config -> Some "Invalid configuration"
    | Invalid_data -> Some "Invalid input data"
    | Error 0 -> Some "General initialization error"
    | Error 1 -> Some "AAC library initialization error"
    | Error 2 -> Some "SBR library initialization error"
    | Error 3 -> Some "Transport library initialization error"
    | Error 4 -> Some "Meta data library initialization error"
    | Error 5 ->
        Some "The encoding process was interrupted by an unexpected error"
    | End_of_file -> Some "End of file reached"
    | _ -> None

  let () = Printexc.register_printer string_of_exception

  type enc
  type info = { input_channels : int; frame_length : int }

  external info : enc -> info = "ocaml_fdkaac_info"

  type t = {
    enc : enc;
    mutable info : info option;
    buffer : Buffer.t;
    chans : int;
  }

  external create : int -> enc = "ocaml_fdkaac_init_enc"

  let create chans =
    { enc = create chans; info = None; buffer = Buffer.create 1024; chans }

  type mpeg2_aac = [ `AAC_LC | `HE_AAC | `HE_AAC_v2 ]
  type mpeg4_aac = [ mpeg2_aac | `AAC_LD | `AAC_ELD ]
  type aot = [ `Mpeg_4 of mpeg4_aac | `Mpeg_2 of mpeg2_aac ]

  let int_of_aot = function
    | `Mpeg_4 `AAC_LC -> 2
    | `Mpeg_4 `HE_AAC -> 5
    | `Mpeg_4 `HE_AAC_v2 -> 29
    | `Mpeg_4 `AAC_LD -> 23
    | `Mpeg_4 `AAC_ELD -> 39
    | `Mpeg_2 `AAC_LC -> 129
    | `Mpeg_2 `HE_AAC -> 132
    | `Mpeg_2 `HE_AAC_v2 -> 156

  let aot_of_int = function
    | 2 -> `Mpeg_4 `AAC_LC
    | 5 -> `Mpeg_4 `HE_AAC
    | 29 -> `Mpeg_4 `HE_AAC_v2
    | 23 -> `Mpeg_4 `AAC_LD
    | 39 -> `Mpeg_4 `AAC_ELD
    | 129 -> `Mpeg_2 `AAC_LC
    | 132 -> `Mpeg_2 `HE_AAC
    | 156 -> `Mpeg_2 `HE_AAC_v2
    | _ -> raise Unsupported_parameter

  type bitrate_mode = [ `Constant | `Full_bitreservoir | `Variable of int ]

  let int_of_bitrate_mode = function
    | `Constant -> 0
    | `Variable vbr -> vbr
    | `Full_bitreservoir -> 8

  let bitrate_mode_of_int = function
    | 0 -> `Constant
    | 8 -> `Full_bitreservoir
    | vbr -> `Variable vbr

  type transmux = [ `Raw | `Adif | `Adts | `Latm | `Latm_out_of_band | `Loas ]

  let int_of_transmux = function
    | `Raw -> 0
    | `Adif -> 1
    | `Adts -> 2
    | `Latm -> 6
    | `Latm_out_of_band -> 7
    | `Loas -> 10

  let transmux_of_int = function
    | 0 -> `Raw
    | 1 -> `Adif
    | 2 -> `Adts
    | 6 -> `Latm
    | 7 -> `Latm_out_of_band
    | 10 -> `Loas
    | _ -> raise Unsupported_parameter

  type param =
    [ `Aot of aot
    | `Bitrate of int
    | `Bitrate_mode of bitrate_mode
    | `Samplerate of int
    | `Sbr_mode of bool
    | `Granule_length of int
    | `Afterburner of bool
    | `Bandwidth of int
    | `Transmux of transmux ]

  type param_name =
    [ `Aot
    | `Bitrate
    | `Bitrate_mode
    | `Samplerate
    | `Sbr_mode
    | `Granule_length
    | `Afterburner
    | `Bandwidth
    | `Transmux ]

  let int_of_param_name = function
    | `Aot -> 0x0100
    | `Bitrate -> 0x0101
    | `Bitrate_mode -> 0x0102
    | `Samplerate -> 0x0103
    | `Sbr_mode -> 0x0104
    | `Granule_length -> 0x0105
    | `Afterburner -> 0x0200
    | `Bandwidth -> 0x0203
    | `Transmux -> 0x0300

  let extract_param = function
    | `Aot x -> (int_of_param_name `Aot, int_of_aot x)
    | `Bitrate x -> (int_of_param_name `Bitrate, x)
    | `Bitrate_mode x -> (int_of_param_name `Bitrate_mode, int_of_bitrate_mode x)
    | `Samplerate x -> (int_of_param_name `Samplerate, x)
    | `Sbr_mode x -> (int_of_param_name `Sbr_mode, if x then 1 else 0)
    | `Granule_length x -> (int_of_param_name `Granule_length, x)
    | `Afterburner x -> (int_of_param_name `Afterburner, if x then 1 else 0)
    | `Bandwidth x -> (int_of_param_name `Bandwidth, x)
    | `Transmux x -> (int_of_param_name `Transmux, int_of_transmux x)

  external set : enc -> int -> int -> unit = "ocaml_fdkaac_set_param"

  let set enc param =
    let p, v = extract_param param in
    set enc.enc p v

  let pack_param = function
    | `Aot, x -> `Aot (aot_of_int x)
    | `Bitrate, x -> `Bitrate x
    | `Bitrate_mode, x -> `Bitrate_mode (bitrate_mode_of_int x)
    | `Samplerate, x -> `Samplerate x
    | `Sbr_mode, x -> `Sbr_mode (x == 1)
    | `Granule_length, x -> `Granule_length x
    | `Afterburner, x -> `Afterburner (x == 1)
    | `Bandwidth, x -> `Bandwidth x
    | `Transmux, x -> `Transmux (transmux_of_int x)

  external get : enc -> int -> int = "ocaml_fdkaac_get_param"

  let get enc param =
    let x = get enc.enc (int_of_param_name param) in
    pack_param (param, x)

  external encode : enc -> string -> int -> int -> string
    = "ocaml_fdkaac_encode"

  (* Drop the first [len] bytes. *)
  let buffer_drop buffer len =
    let size = Buffer.length buffer in
    if len = size then Buffer.clear buffer
    else begin
      let tmp = Buffer.sub buffer len (size - len) in
      Buffer.clear buffer;
      Buffer.add_string buffer tmp
    end

  let rec encode_frames enc len pos out =
    if len <= Buffer.length enc.buffer - pos then begin
      let data = Buffer.sub enc.buffer pos len in
      Buffer.add_string out (encode enc.enc data 0 len);
      encode_frames enc len (pos + len) out
    end
    else if 0 < pos then buffer_drop enc.buffer pos

  let get_info enc =
    match enc.info with
      | Some info -> info
      | None ->
          let info = info enc.enc in
          enc.info <- Some info;
          info

  let chunk_length enc =
    let info = get_info enc in
    2 * info.input_channels * info.frame_length

  let encode enc data ofs len =
    if String.length data - ofs < len then raise Invalid_data;
    Buffer.add_substring enc.buffer data ofs len;
    let ret = Buffer.create 1024 in
    encode_frames enc (chunk_length enc) 0 ret;
    Buffer.contents ret

  external flush : enc -> string = "ocaml_fdkaac_flush"

  let flush enc =
    let buf = Buffer.create 1024 in
    let len = chunk_length enc in
    encode_frames enc len 0 buf;
    let rem = Buffer.length enc.buffer in
    if 0 < rem then begin
      Buffer.add_string enc.buffer (String.make (len - rem) (Char.chr 0));
      encode_frames enc len 0 buf
    end;
    try
      while true do
        Buffer.add_string buf (flush enc.enc)
      done;
      assert false
    with Unsupported_parameter | End_of_file -> Buffer.contents buf
end
